include/mbedtls/rsa.h

```diff
@@ -74,6 +74,169 @@ extern "C" {
 #endif
 
 /**
+ * Helper functions for RSA-related operations on MPI's.
+ */
+
+/**
+ * \brief          Compute RSA prime moduli P, Q from public modulus N=PQ
+ *                 and a pair of private and public key.
+ *
+ * \note           This is a 'static' helper function not operating on
+ *                 an RSA context. Alternative implementations need not
+ *                 overwrite it.
+ *
+ * \param N        RSA modulus N = PQ, with P, Q to be found
+ * \param D        RSA private exponent
+ * \param E        RSA public exponent
+ * \param f_rng    PRNG to be used for randomization, or NULL
+ * \param p_rng    PRNG context for f_rng, or NULL
```

Comment:
@gilles-peskine-arm That's a very good question... the number of iterations needed without a PRNG depends on chains of quadratic residues, and such can get arbitrarily long. So there's no static bound on the number of iterations - but of course, no such bound would exist when a PRNG was used, either.

I don't have an indication that the heuristic would be less efficient in average than one using a PRNG, or one starting the iteration at a different base than `sqrt(N)` (the current documentation wrongly says it starts around `N/2`).

I've run a few tests with different approaches:
1. Random choice of candidates
2. Consecutive candidates `2, 3, 4, 5, ...`
3. Prime candidates `2, 3, 5, 7, 11, ...`
4. Consecutive candidates `B, B+1, B+2, ...` with `B ~ sqrt(N)`. 
5. Prime candidates `3, 5, 7, 11, ...`
6. Prime candidates `2, 3, 5, 7, 11, ...`, but omit `2` if `N == 1 mod 8`.

Here are the results:

```
-------------------------------
2, 3, 4, 5, 6, 7, ...

Total runs: 10000
Attempts:   16763
Tests with 01 attempts: 07142 (71%)
Tests with 02 attempts: 01850 (18%)
Tests with 03 attempts: 00000 (00%)
Tests with 04 attempts: 00560 (05%)
Tests with 05 attempts: 00037 (00%)
Tests with 06 attempts: 00245 (02%)
Tests with 07 attempts: 00000 (00%)
Tests with 08 attempts: 00000 (00%)
Tests with 09 attempts: 00006 (00%)
Tests with 10 attempts: 00079 (00%)
Tests with 11 attempts: 00000 (00%)
Tests with 12 attempts: 00047 (00%)
Tests with 13 attempts: 00001 (00%)
Tests with 14 attempts: 00000 (00%)
Tests with 15 attempts: 00000 (00%)
Tests with 16 attempts: 00018 (00%)
Tests with 17 attempts: 00000 (00%)
Tests with 18 attempts: 00008 (00%)
Tests with 19 attempts: 00000 (00%)
Tests with 20 attempts: 00000 (00%)
Tests with 21 attempts: 00001 (00%)
Tests with 22 attempts: 00003 (00%)
Tests with 23 attempts: 00000 (00%)
Tests with 24 attempts: 00000 (00%)
Tests with 25 attempts: 00000 (00%)
Tests with 26 attempts: 00000 (00%)
Tests with 27 attempts: 00000 (00%)
Tests with 28 attempts: 00002 (00%)
Tests with 29 attempts: 00000 (00%)
Tests with 30 attempts: 00001 (00%)

-----------------------------------
2, 3, 5, 7, 11, ...

Total runs: 10000
Attempts:   14660
Tests with 01 attempts: 07181 (71%)
Tests with 02 attempts: 01820 (18%)
Tests with 03 attempts: 00557 (05%)
Tests with 04 attempts: 00239 (02%)
Tests with 05 attempts: 00097 (00%)
Tests with 06 attempts: 00058 (00%)
Tests with 07 attempts: 00023 (00%)
Tests with 08 attempts: 00016 (00%)
Tests with 09 attempts: 00002 (00%)
Tests with 10 attempts: 00005 (00%)
Tests with 11 attempts: 00002 (00%)
Tests with 12 attempts: 00000 (00%)
Tests with 13 attempts: 00000 (00%)
Tests with 14 attempts: 00000 (00%)
Tests with 15 attempts: 00000 (00%)
Tests with 16 attempts: 00000 (00%)
Tests with 17 attempts: 00000 (00%)
Tests with 18 attempts: 00000 (00%)
Tests with 19 attempts: 00000 (00%)
Tests with 20 attempts: 00000 (00%)
Tests with 21 attempts: 00000 (00%)
Tests with 22 attempts: 00000 (00%)
Tests with 23 attempts: 00000 (00%)
Tests with 24 attempts: 00000 (00%)
Tests with 25 attempts: 00000 (00%)
Tests with 26 attempts: 00000 (00%)
Tests with 27 attempts: 00000 (00%)
Tests with 28 attempts: 00000 (00%)
Tests with 29 attempts: 00000 (00%)
Tests with 30 attempts: 00000 (00%)

--------------------
RANDOM

Total runs: 10000
Attempts:   14224
Tests with 01 attempts: 07445 (74%)
Tests with 02 attempts: 01623 (16%)
Tests with 03 attempts: 00549 (05%)
Tests with 04 attempts: 00188 (01%)
Tests with 05 attempts: 00110 (01%)
Tests with 06 attempts: 00049 (00%)
Tests with 07 attempts: 00016 (00%)
Tests with 08 attempts: 00012 (00%)
Tests with 09 attempts: 00002 (00%)
Tests with 10 attempts: 00004 (00%)
Tests with 11 attempts: 00001 (00%)
Tests with 12 attempts: 00000 (00%)
Tests with 13 attempts: 00001 (00%)
Tests with 14 attempts: 00000 (00%)
Tests with 15 attempts: 00000 (00%)
Tests with 16 attempts: 00000 (00%)
Tests with 17 attempts: 00000 (00%)
Tests with 18 attempts: 00000 (00%)
Tests with 19 attempts: 00000 (00%)
Tests with 20 attempts: 00000 (00%)
Tests with 21 attempts: 00000 (00%)
Tests with 22 attempts: 00000 (00%)
Tests with 23 attempts: 00000 (00%)
Tests with 24 attempts: 00000 (00%)
Tests with 25 attempts: 00000 (00%)
Tests with 26 attempts: 00000 (00%)
Tests with 27 attempts: 00000 (00%)
Tests with 28 attempts: 00000 (00%)
Tests with 29 attempts: 00000 (00%)
Tests with 30 attempts: 00000 (00%)

-----------------------------------
sqrt(N) ~ B, B+1, B+2, ...

Total runs: 10000
Attempts:   14273
Tests with 01 attempts: 07357 (73%)
Tests with 02 attempts: 01701 (17%)
Tests with 03 attempts: 00565 (05%)
Tests with 04 attempts: 00215 (02%)
Tests with 05 attempts: 00082 (00%)
Tests with 06 attempts: 00041 (00%)
Tests with 07 attempts: 00024 (00%)
Tests with 08 attempts: 00007 (00%)
Tests with 09 attempts: 00004 (00%)
Tests with 10 attempts: 00001 (00%)
Tests with 11 attempts: 00003 (00%)
Tests with 12 attempts: 00000 (00%)
Tests with 13 attempts: 00000 (00%)
Tests with 14 attempts: 00000 (00%)
Tests with 15 attempts: 00000 (00%)
Tests with 16 attempts: 00000 (00%)
Tests with 17 attempts: 00000 (00%)
Tests with 18 attempts: 00000 (00%)
Tests with 19 attempts: 00000 (00%)
Tests with 20 attempts: 00000 (00%)
Tests with 21 attempts: 00000 (00%)
Tests with 22 attempts: 00000 (00%)
Tests with 23 attempts: 00000 (00%)
Tests with 24 attempts: 00000 (00%)
Tests with 25 attempts: 00000 (00%)
Tests with 26 attempts: 00000 (00%)
Tests with 27 attempts: 00000 (00%)
Tests with 28 attempts: 00000 (00%)
Tests with 29 attempts: 00000 (00%)
Tests with 30 attempts: 00000 (00%)

-----------------------------------
3, 5, 7, 11, 13, ...

Total runs: 10000
Attempts:   14288
Tests with 01 attempts: 07388 (73%)
Tests with 02 attempts: 01668 (16%)
Tests with 03 attempts: 00548 (05%)
Tests with 04 attempts: 00238 (02%)
Tests with 05 attempts: 00069 (00%)
Tests with 06 attempts: 00042 (00%)
Tests with 07 attempts: 00026 (00%)
Tests with 08 attempts: 00007 (00%)
Tests with 09 attempts: 00009 (00%)
Tests with 10 attempts: 00003 (00%)
Tests with 11 attempts: 00002 (00%)
Tests with 12 attempts: 00000 (00%)
Tests with 13 attempts: 00000 (00%)
Tests with 14 attempts: 00000 (00%)
Tests with 15 attempts: 00000 (00%)
Tests with 16 attempts: 00000 (00%)
Tests with 17 attempts: 00000 (00%)
Tests with 18 attempts: 00000 (00%)
Tests with 19 attempts: 00000 (00%)
Tests with 20 attempts: 00000 (00%)
Tests with 21 attempts: 00000 (00%)
Tests with 22 attempts: 00000 (00%)
Tests with 23 attempts: 00000 (00%)
Tests with 24 attempts: 00000 (00%)
Tests with 25 attempts: 00000 (00%)
Tests with 26 attempts: 00000 (00%)
Tests with 27 attempts: 00000 (00%)
Tests with 28 attempts: 00000 (00%)
Tests with 29 attempts: 00000 (00%)
Tests with 30 attempts: 00000 (00%)

-----------------------------------
2 (if N % 8 != 1), 3, 5, 7, 11, ...

Total runs: 10000
Attempts:   12644
Tests with 01 attempts: 08218 (82%)
Tests with 02 attempts: 01280 (12%)
Tests with 03 attempts: 00305 (03%)
Tests with 04 attempts: 00107 (01%)
Tests with 05 attempts: 00052 (00%)
Tests with 06 attempts: 00021 (00%)
Tests with 07 attempts: 00009 (00%)
Tests with 08 attempts: 00003 (00%)
Tests with 09 attempts: 00002 (00%)
Tests with 10 attempts: 00001 (00%)
Tests with 11 attempts: 00002 (00%)
Tests with 12 attempts: 00000 (00%)
Tests with 13 attempts: 00000 (00%)
Tests with 14 attempts: 00000 (00%)
Tests with 15 attempts: 00000 (00%)
Tests with 16 attempts: 00000 (00%)
Tests with 17 attempts: 00000 (00%)
Tests with 18 attempts: 00000 (00%)
Tests with 19 attempts: 00000 (00%)
Tests with 20 attempts: 00000 (00%)
Tests with 21 attempts: 00000 (00%)
Tests with 22 attempts: 00000 (00%)
Tests with 23 attempts: 00000 (00%)
Tests with 24 attempts: 00000 (00%)
Tests with 25 attempts: 00000 (00%)
Tests with 26 attempts: 00000 (00%)
Tests with 27 attempts: 00000 (00%)
Tests with 28 attempts: 00000 (00%)
Tests with 29 attempts: 00000 (00%)
Tests with 30 attempts: 00000 (00%)
```

Except from a heuristical argument that it is unlikely (though not impossible, as the stats show) that the first succeeding candidate when trying `2, 3, 4, ...` is a non-prime, I haven't yet backed up these results with theoretical understanding. I'll try to give it some more thought in the spare time.

Based on the statistics only, I'd currently vote for removing the PRNG from the interface and follow approach (6), i.e. picking small primes as the candidates and omitting `2` if `N == 1 (mod 8)`.

Also, note that in contrast to e.g. Miller-Rabin, bad candidate choices cannot lead to a false positive. If the static choices happen to be bad, the code just runs longer.

Reply:
@hanno-arm This is just a randomised mathematical algorithm, no crypto involved, you don't need the callback here. The built-in congruential generator is perfectly fine for this purpose.

Reply:
@yanesca Yes! I'll drop the CSPRNG from the interface and will either go with the congruential generator you mentioned or with the static approach (6) mentioned above. In any case, it's just an implementation-choice that can be changed at any time, while the interface is something we have to live with.

Reply:
@yanesca while I agree that in principle the libc's `rand()` function is fine for this purpose, I'd still advocate against using it. Many tools (coverity, etc) will flag any use of this function (rightfully IMO, the tool can'ts know that particular use is OK) and we then have to spend time manually ignoring it (including in reports from users etc).

In that case Hanno intended to use a static approach anyway, so it doesn't matter, but I thought it might be worth mentioning if similar cases arise later.

Reply:
Another reason not to use `rand` is that we target freestanding environments that might not have a full libc.